[reland] Refactor ToggleButtons (remove RawMaterialButton)

TahaTesser opened the pull request:
reland https://github.com/flutter/flutter/pull/99493

fixes https://github.com/flutter/flutter/issues/99085
fixes https://github.com/flutter/flutter/issues/97302


| Before | After |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/163001719-241316c8-8e1a-494b-8b03-80db2d0ab254.jpg"  /> | <img src="https://user-images.githubusercontent.com/48603081/164722763-9b0374a4-8460-4a4d-ac84-d83690ebc48d.png" /> |





Individual buttons are expected to take up all the available space but that restrict providing `_inputPadding` to individual buttons, children can't have smaller size while keeping padding (this requires individual buttons to be wrapped with `Center`) so individual buttons only take the space they need and leave room for input padding. 

But buttons are needed to take all available when you have children of different sizes then all button sizes should match the tallest widget hence adding the `expandChildren` property. This property will allow children to expand and take the full space.

<details> 
<summary>minimal code sample</summary> 

```dart
import 'package:flutter/material.dart';

class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  _HomeState createState() => _HomeState();
}

class _HomeState extends State<Home> {
  int selectedTabIndex = 0;

  @override
  Widget build(BuildContext context) {
    final ColorScheme colorScheme = Theme.of(context).colorScheme;
    return Scaffold(
      body: Center(
        child: Container(
          color: Colors.green.withOpacity(0.15),
          child: ToggleButtons(
            direction: Axis.horizontal,
            tapTargetSize: MaterialTapTargetSize.padded,
            isSelected: [
              selectedTabIndex == 0,
              selectedTabIndex == 1,
            ],
            onPressed: (int index) {
              setState(() {
                selectedTabIndex = index;
              });
            },
            borderRadius: const BorderRadius.all(Radius.circular(8)),
            selectedBorderColor: colorScheme.primary,
            selectedColor: colorScheme.primary,
            constraints: const BoxConstraints.tightFor(
              width: 86,
              height: 32,
            ),
            // expandedChildren: true,
            children: const <Widget>[
              Text('Locks'),
              Text('Unlocks'),
            ],
          ),
        ),
      ),
    );
  }
}

void main() {
  runApp(const MaterialApp(home: Home()));
}

``` 
	
</details>

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

TahaTesser commented on packages/flutter/test/material/toggle_buttons_test.dart:
```diff
@@ -1910,4 +1912,103 @@ void main() {
+    expect(thirdRect.height, 48.0);
+  });
+
+  testWidgets('Toggle buttons constraints size does not affect minimum input padding', (WidgetTester tester) async {
```
This new tests checks for button size and semantic node size
button's = 32.0
semantic node's height 48.0

cc: @HansMuller

math1man commented:
Nice! Looks great!

HansMuller commented:
I'm afraid I couldn't quite parse the rationale for `expandChildren`.  What's is the scenario where `expandChildren` is neccessary?

TahaTesser commented:
@HansMuller 

Here in the code, toggle buttons are stretched to take all the available space to match the tallest widget

https://github.com/flutter/flutter/blob/5fb74db1a9cb8c57dbf58c1611159e4c31d2d43d/packages/flutter/lib/src/material/toggle_buttons.dart#L701-L716

For this test 
https://github.com/flutter/flutter/blob/5fb74db1a9cb8c57dbf58c1611159e4c31d2d43d/packages/flutter/test/material/toggle_buttons_test.dart#L1177-L1188

This makes toggle buttons take all the available space, and stretch and input paddings are not possible as a result.

Here on the left,  buttons with `constraints: const BoxConstraints.tightFor(width: 86, height: 32)` are stretch to take `48.0` height (default height of `TextButton`)

On the right, all buttons are stretched to match the tallest widget with various size constraints. Code from `Height of segmented control is determined by the tallest widget` test.

| Buttons with `32.0` height constraint (stretched to `48.0`)  | Buttons with various height constraints  |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/163122553-8fbe4431-7206-4fd8-880f-1be20550a8fd.png" /> | <img src="https://user-images.githubusercontent.com/48603081/163121896-3cabc4d3-c32b-45e0-a81f-f9eb641b3db1.png" />




We need the buttons to take less than `48.0` as per specs (existing design), I added the `Center` widget so buttons are not stretched and they are allowed to have the actual size they have constrained to. 


| Buttons with `32.0`  | Buttons with various height constraints  |
| --------------- | --------------- |
| <img src="https://user-images.githubusercontent.com/48603081/163121888-4b10ef0b-c96f-4578-b7f0-037135d75e4b.png" /> | <img src="https://user-images.githubusercontent.com/48603081/163123141-ee24da58-cda3-4036-b083-28a34f256c68.png" />


We need buttons not to stretch and leave room for padding when height but as you can see on the right, all buttons are constrained when test expect the buttons to be stretched to match the tallest widget height, hence I added  `expandChildren` which allows buttons to expand when needed. This might not be the best solution, if you have a better idea, please let me know, appreciate it

Piinks commented:
I'm testing this out now @TahaTesser, thanks for your patience. It looks like there are merge conflicts currently though, just FYI. :)

TahaTesser committed:
[reland] Refactor ToggleButtons (remove RawMaterialButton)

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -416,6 +418,11 @@ class ToggleButtons extends StatelessWidget {
+  /// Whether to expand the children to take height of the tallest widget.
+  ///
+  /// Defaults to false.
+  final bool expandChildren;
```
I think I found a way for this to not be necessary. I played with this a bit today and found that all existing and new tests pass if we base this on whether or not constraints are provided, instead of using this flag.

Piinks replied:
```suggestion
```
Without this, the same can be achieved with the following changes

TahaTesser replied:
This is brilliant, I didn't think about nulling max constraints.

Piinks replied:
Thanks! I can't take too much credit. I migrated all of the customers inside of Google off of FlatButton, RaisedButton and OutlineButton. It took a while to figure out! :)

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,33 +674,127 @@ class ToggleButtons extends StatelessWidget {
+      final TextStyle currentTextStyle = textStyle
+        ?? toggleButtonsTheme.textStyle
+        ?? theme.textTheme.bodyText2!;
+      final BoxConstraints currentConstraints = constraints
+        ?? toggleButtonsTheme.constraints
+        ?? const BoxConstraints(
+             minWidth: kMinInteractiveDimension,
+             minHeight: kMinInteractiveDimension,
+           );
```
```suggestion
final BoxConstraints? currentConstraints = constraints
          ?? toggleButtonsTheme.constraints;
      final Size minimumSize = currentConstraints == null
          ? const Size.square(kMinInteractiveDimension)
          : Size(currentConstraints.minWidth, currentConstraints.minHeight);
      final Size? maximumSize = currentConstraints == null
          ? null
          : Size(currentConstraints.maxWidth, currentConstraints.maxHeight);
```

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,33 +674,127 @@ class ToggleButtons extends StatelessWidget {
+             minHeight: kMinInteractiveDimension,
+           );
+
+      final Size minSize;
```
Change this to minPaddingSize or some such for clarity
```suggestion
      final Size minPaddingSize;
```

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,33 +674,127 @@ class ToggleButtons extends StatelessWidget {
+                color: currentColor,
+              )),
+              padding: MaterialStateProperty.all<EdgeInsetsGeometry>(EdgeInsets.zero),
+              minimumSize: MaterialStateProperty.all<Size?>(
+                Size(currentConstraints.minWidth, currentConstraints.minHeight)),
+              maximumSize: MaterialStateProperty.all<Size?>(
+                Size(currentConstraints.maxWidth, currentConstraints.maxHeight)),
```
```suggestion
              minimumSize: MaterialStateProperty.all<Size?>(minimumSize),
              maximumSize: MaterialStateProperty.all<Size?>(maximumSize),
```

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,33 +674,127 @@ class ToggleButtons extends StatelessWidget {
+        ),
+      );
+
+      if (!expandChildren) {
```
```suggestion
      if (currentConstraints != null) {
```

Piinks reviewed (commented):
I think that should work, please take a look and let me know what you think. And as always, thank you!

TahaTesser committed:
Fixes fron review and clean up

TahaTesser commented:
I also made another minor change, since `result` widget is not wrapped in input padding, We don't need to return it from the widget every time from the if-else statement, instead, I am returning the default configuration if the axis direction is the default.


[hhttps://github.com/flutter/flutter/pull/101760/files#diff-153bdbb7a6739dd071bbb7b8230e3884dcb6cedbc21a78ef82c0e6a8166bb220R795-R811](https://github.com/flutter/flutter/pull/101760/files#diff-153bdbb7a6739dd071bbb7b8230e3884dcb6cedbc21a78ef82c0e6a8166bb220R805-R811)

Piinks commented:
Cool! I will take another look and run it against the internal test that caused it to be reverted just to be sure.

Piinks commented:
It looks like the previous test failure is fixed now internally, but we introduced some new ones here. I'll update once I know what broke these other tests.

Piinks commented:
Still working on resolving the test failures. I am hoping to have them sorted today, a customer has informed us that #97302 is a potential release blocker and this will fix it.

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,64 +668,147 @@ class ToggleButtons extends StatelessWidget {
+        case MaterialTapTargetSize.padded:
+          if (direction == Axis.horizontal) {
+            minPaddingSize = const Size(
+              kMinInteractiveDimension,
```
```suggestion
              0.0,
```

TahaTesser replied:
My understanding was input padding height needed to match 48.0 when regardless of the custom [constraints](https://api.flutter.dev/flutter/material/ToggleButtons/constraints.html)  but I was probably mistaken and the accessibility green window should take input padding space.

![image](https://user-images.githubusercontent.com/48603081/164681996-2732beac-e5ec-4e70-a17e-b1d1e9b92767.png)


To be double sure, I just checked the native reference and it does look like input padding is not supposed to be higher than the toggle button.

![2022-04-22 12 41 08](https://user-images.githubusercontent.com/48603081/164682020-3dcfc39d-2fd7-4994-9ea4-30729a68bf7a.jpg)



This is fixed by this suggestion provided here.


![2022-04-22 12 41 03](https://user-images.githubusercontent.com/48603081/164682029-26aa0371-ef3d-418f-ad9f-fa217b177b24.jpg)

math1man replied:
The input padding is supposed to be higher than the toggle button. Even in the screenshots provided here, you can see the difference. In the middle screenshot, the green rectangle is taller than the "everyone" button, while in the bottom screenshot the green rectangle is fully contained within the "locks" button. Here is a screenshot highlighting the exact bounds of the "everyone" button:

![image](https://user-images.githubusercontent.com/4040946/164713826-4b71a33e-20e8-4451-804b-336ab08c14c9.png)

Note the slight difference in height between the rendering and the actual bounds.

TahaTesser replied:
Thank for highlighting this, the semantic focus window indeed looks slightly taller but not `48.0` from the looks of it.

If we set `kMinInteractiveDimension` , it would rather look like this. 
![image](https://user-images.githubusercontent.com/48603081/164718368-4c220302-4bc6-492f-b3c2-583be0092047.png)


Than this 

![image](https://user-images.githubusercontent.com/48603081/164718390-ff8f403f-39d6-44b6-bc00-655ab34dbfa4.png)


 Could it be that the focus window is wrapping like an outer border in the native app and wrapping like an inner border in Flutter?.

You can see my updated screenshot, green border looks like an inner border? and the native focus window looks like an outer border. 

Since it is not a standard height like `48.0` so what height it should be? or vertical padding like `4.0` something? I don't see such a design in other widgets. 


`Chip` has such padding and it is `kMinInteractiveDimension`.
https://github.com/flutter/flutter/blob/fde7622f7e0ab915ce19bd49a8ead5707141df6a/packages/flutter/lib/src/material/chip.dart#L1241-L1252



![2022-04-22 15 53 12](https://user-images.githubusercontent.com/48603081/164718226-61f3bab1-0102-422f-ad30-386fc49aa77a.jpg)
![2022-04-22 15 53 16](https://user-images.githubusercontent.com/48603081/164718231-857a7558-57cf-45b1-b348-cfac1bc7aa2c.jpg)

Piinks replied:
I tested this against your project's semantics tests yesterday @math1man, and I believe this produces the desired behavior. I will send you the report. I know we originally thought we solved this in https://github.com/flutter/flutter/issues/93249, so I want o make sure we get it right this time.

Piinks replied:
@TahaTesser l#708 is the width, isn't it? It looks like l#709 should be `kMinInteractiveDimension` for the height.

TahaTesser replied:
My apologies just fixed this.

Piinks replied:
No worries at all! :)

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,64 +668,147 @@ class ToggleButtons extends StatelessWidget {
+        button = Center(child: button);
+      }
+
+      return Semantics(
+        container: true,
+        button: true,
+        enabled: onPressed != null,
+        child: _InputPadding(
+          minSize: minPaddingSize,
+          direction: direction,
+          child: Semantics(
+            excludeSemantics: true,
+            child: button,
+          ),
```
```suggestion
return MergeSemantics(
        child: Semantics(
          container: true,
          enabled: onPressed != null,
          child: _InputPadding(
            minSize: minPaddingSize,
            direction: direction,
            child: button,
          ),
```

Piinks reviewed (dismissed):
These two adjustments should have this sorted. Can you apply these updates and I'll just double check once more? I think the Google testing shard is fixed now too if you want to pull updates from master. Thanks!

TahaTesser committed:
Updates from review

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,64 +668,149 @@ class ToggleButtons extends StatelessWidget {
+          if (direction == Axis.horizontal) {
+            minPaddingSize = const Size(
+              kMinInteractiveDimension,
+              0.0,
```
```suggestion
              kMinInteractiveDimension,
```

TahaTesser replied:
Done.

TahaTesser committed:
Update padding size

Piinks commented:
When you have a chance @TahaTesser can you update the before/after image in your OP? I think the mix up with x and y is what caused the unexpected image this morning. Thank you!

TahaTesser commented:
> When you have a chance @TahaTesser can you update the before/after image in your OP? I think the mix up with x and y is what caused the unexpected image this morning. Thank you!

Restored :)

Piinks commented on packages/flutter/lib/src/material/toggle_buttons.dart:
```diff
@@ -667,64 +668,149 @@ class ToggleButtons extends StatelessWidget {
+          child: _InputPadding(
+            minSize: minPaddingSize,
+            direction: direction,
+            child: Semantics(
+              excludeSemantics: true,
+              child: button,
```
@TahaTesser this is still off, this additional Semantics widget should not be here. Excluding the semantics of the child leaves it blank without any context.

Piinks replied:
See comment here: https://github.com/flutter/flutter/pull/101760#discussion_r855654325

TahaTesser committed:
Update sematics and semantics test

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/165597540-31895fc5-5a9f-4c7a-8e7e-df499486d04e.png)
